### PR TITLE
[Mobile Payments] Present incomplete store address error to user

### LINKED ISSUE #5088: [Mobile Payments] Present to user failure to fetch location when connecting to a reader
Right now we handle golden path only. We should present to the user a message when we are unable to fetch a location ID for their reader. Disambiguating from other connection failures should help with debugging connection failures.

Needs design?

### PR DESCRIPTION
**Pull request by joshheald**
Closes #5088

### Description
This change will present an error to the user when they need to update their store address in order to connect to a reader. In a further PR, it will include a button with a link to the WP-Admin page where they can make the change.

### Testing instructions

1. With a store set up for In-Person Payments, go to `WP-Admin > WooCommerce > Settings` and delete the contents of `Address line 1`
2. In the app, go to `⚙️ > In-Person Payments > Manage Card Reader`
3. Search for and connect to a reader (or simulated reader)
4. Observe that the error message is shown telling the merchant to update their address.
5. Update the address in WP-Admin to include `Address line 1` again
6. Tap retry, and connect to the reader again.

Unit tests will follow in a future PR. The retry action could use some polish; e.g. just retrying connection to the exact reader which failed would be much nicer than starting a search again... but that can be an iteration on this for after we send the user to the `adminUrl` or something.

### Screenshots

https://user-images.githubusercontent.com/2472348/143595235-441c848e-44e8-41c9-a1da-0646599855bb.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by joshheald** [comment]
@ctarda I'd appreciate your thoughts on this as I'm starting to agonise over this. In particular, do you think I'm putting too much into `UnderlyingError+Stripe.swift` with it knowing about `DotcomError` and depending on `Networking`. 

I'm not a massive fan of that dependency, but it does seem like the best place for us to figure out that we've got this error, if we want to present it wherever we need to.

**Comment by joshheald** [comment]
Hmm... tests on CI aren't happy about the dependency either!

File: Hardware/Hardware/CardReader/StripeCardReader/UnderlyingError+Stripe.swift
Comment (ctarda) [inline Hardware/Hardware/CardReader/StripeCardReader/UnderlyingError+Stripe.swift]: @joshheald As you mentioned in the comments, it seems like we might mixing concerns a little bit, with leads to tangle the dependency tree a little bit as well.

It seems to me like what we want is to pass up to the UI an error that is not actually thrown by the Terminal SDK, but that happens when a network request is failing. 

In one sentence: we want to pass an error to the user when a network request is failing. That request is the one that fetches the location, and is implemented by WCPayRemote.

That request fetches the location id, and what we want is to:
1. Stop the attempt to connect to the reader
2. Show a user facing error.

To me, the key is point #1. We want to stop the connection to a reader when we can not get a location identifier. That, to me, is business logic.

That's all to say that I think we should move the logic out of Hardware (which doesn't do any other networking) at least one layer up.

UnderlyigError conforms to Error, so we could do a bit of good old fashioned polymorphism. WCPayTokenProvider is the entity that encapsulates the business logic of fetching the location id. At this point, it is passing up whatever error is coming from the networking layer. 

I would explore declaring a new error alongside it (it could be an inner type to WCPayTokenProvider or it could be just a public entity declared next to it), so that WCPayTokenProvider.fetchDefaultLocationID actually throws one instance of this new error.

If everything else works as expected, this new error can be passed to CardReaderServiceError as an underlying error in when captured in StripeCardReaderService.getBluetoothConfiguration (this might or might not require adding a new case to CardReaderServiceError)

From that on, the error should be propagated up to the UI without other intervention. It might be necessary to add some logic to the UI as well to make sure that the payment intent is cancelled if connection to the reader fails.

But, in a nutshell, I would explore to:
1. Declare a new error type that would be initialized by WCPayTokenProvider when the request to get the location id fails.
2. Pass that new error in the completion block of fetchDefaultLocationID, instead of the Dotcom error that is coming from the remote service.
3. Take it from there 😄

File: Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift
Comment (joshheald) [inline Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift]: I'm leaving this slightly confusingly named function (that doesn't necessarily make an `UnderlyingError`) because I think that the next steps are to make `UnderlyingError` an abstract parent of both `StripeUnderlyingError` (or something) and `CardReaderConfigError`

Comment (ctarda) [reply Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift]: I would prefer to stay away from creating a hierarchy of errors. All of them implement Error after all, so conforming to Error seems like is all we might need.

I do agree this function name could be less confusing. What about extracting it to a factory in UnderlyingError? Something like

```
extension UnderlyingError {
   init(with configurationError: CardReaderConfigurationError) {
   }
}
```

That way, I think, we could also avoid the downcast to CardReaderConfigError as well.

Comment (ctarda) [reply Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift]: Actually, I am not sure if it's actually possible to type the parameter as CardReaderConfigurationError, so the downcast might not be possible to avoid

Comment (joshheald) [reply Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift]: Yeah I tried to make the hierarchy and quickly noped out of that, so I'm glad you don't love that idea!

I'll work on the naming. I could maybe remove this method by using a failable initializer in the extension you suggest.

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (joshheald) [inline Hardware/Hardware/CardReader/CardReaderServiceError.swift]: I'd read `UnderlyingError` as being more bound to Stripe, but really that's just the extension with Stripe in the name, so adding this makes sense and avoids some of the polymorphism gymnastics I was having to do. Thanks!

Comment (joshheald) [reply Hardware/Hardware/CardReader/CardReaderServiceError.swift]: I think this could do with also being called `incompleteStoreAddress`, since we want to surface that error to the merchant, and will need to include the `adminUrl` associated value as well.

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (joshheald) [inline Hardware/Hardware/CardReader/CardReaderServiceError.swift]: It's convenient to fall back to the `+Stripe` extension here, which provides default behaviour of mapping errors it can't parse to `.internalServerError`... but it does seem like a bit of a side effect.

Two options come to mind:
1. Make this a failable init, and then at the call site use the Stripe init where it's appropriate.
2. Add a `.unknown` case to the `UnderlyingError` and use that, rather than always falling back to `.internalServerError`

I'm going to try option 1, because I think it'll be more explicit and help us see any consequences of the change.

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (joshheald) [inline Hardware/Hardware/CardReader/CardReaderServiceError.swift]: Not totally sure we need this for a single case? We might need to add `.storePostcodeMissing` for #5320, in which case it could be helpful, but I could also see that error being collapsed into this one. Regardless, we'll need to unwrap and downcast at the call site so that we can get the `adminUrl` out when that's included

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift]: I used that name first too, and didn't like it because it's so similar to the enclosing method name. 

Did you want to inline the error creation here for a particular reason? I quite liked extracting it to another method which then differentiated between the default alert and the missing address alert. I'm going to go back to that for now, and try to name stuff better, as the previous code was handling both alert cases.

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift]: Yeah, agreed, planning to do that in a later commit/PR

**Review by joshheald — COMMENTED** [review]
@ctarda thanks for doing so much on it! I've added a review as a more convenient way for us to discuss, and as notes for myself as I work on these changes. Chime in with any thoughts 😊

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (ctarda) [reply Hardware/Hardware/CardReader/CardReaderServiceError.swift]: Absolutely! I am terrible at naming things

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (ctarda) [reply Hardware/Hardware/CardReader/CardReaderServiceError.swift]: I would lean more towards option 2, to be honest. The way we would not need to deal with special cases at the call site

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift
Comment (ctarda) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift]: I have no preference here, to be honest. This looked slightly less verbose to me, and that's why I went with it 😄

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (ctarda) [reply Hardware/Hardware/CardReader/CardReaderServiceError.swift]: Maybe a more generic name like `isLocationError`? It seems like marking these special cases clearly simplifies the client code a lot.

File: Hardware/Hardware/CardReader/CardReaderServiceError.swift
Comment (joshheald) [reply Hardware/Hardware/CardReader/CardReaderServiceError.swift]: To follow up from a slack chat; I gave the `.unknown` case a go here and it didn't really add anything. `.internalServiceError` is basically the unknown case already, it's intended as a catch all, and that's how we use it. 

It is quite close to internalServ**er**Error, which might lead to some confusion, so we could consider renaming it.

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalConnectionFailedUpdateAddress.swift
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalConnectionFailedUpdateAddress.swift:48]: Could we be more specific? i.e. "update" seems vague. Could we use the word "correct" or "fully set" or something instead? cc @adamzelinski

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift
Comment (allendav) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift:587]: Instead of MissingAddress, would it be more appropriate to say InvalidStoreAddress or BadStoreAddress?

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift
Comment (allendav) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift:31]: Instead of continue, since it looks like we need to "retry" it, could we say "retrySearch" ?

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift
Comment (allendav) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift:169]: Same comment as above

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlertsProvider.swift
Comment (allendav) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlertsProvider.swift:51]: Instead of MissingAddress, would it be more appropriate to say InvalidStoreAddress or BadStoreAddress?

File: WooCommerce/WooCommerceTests/Mocks/MockCardReaderSettingsAlerts.swift
Comment (allendav) [inline WooCommerce/WooCommerceTests/Mocks/MockCardReaderSettingsAlerts.swift:87]: Same comments as above re missing and continue

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (allendav) [inline Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:245]: Do we need a separate error (CardReaderConfigError) apart from the CardReaderServiceError enumeration you've added for bad address above?

Since the CardPresentPaymentStore manages all access to the CardReaderService, it would seem appropriate to leave the function signatures as Error and construct and return a CardReaderService.incompleteStoreAddress in the event the DotcomError has store_address_is_incomplete, but otherwise forward the Error intact from the remote?

**Review by allendav — COMMENTED** [review]
I'm curious of your thoughts, especially on https://github.com/woocommerce/woocommerce-ios/pull/5515/files#r759598494

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift
Comment (allendav) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift:587]: In light of #5560, which I noticed after I reviewed this PR, I'd opt for "Incomplete" instead of "Missing"

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlertsProvider.swift
Comment (allendav) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlertsProvider.swift:51]: In light of #5560, which I noticed after I reviewed this PR, I'd opt for "Incomplete" instead of "Missing"

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift:587]: Because there's a string of PRs to merge here, to avoid messy merges I'm going to open issues for minor changes and address them next, after merging all PRs. If there's anything fundamental, I won't merge until fixing, but from a skim it doesn't look like there is. See #5567

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift:31]: This is done in #5552 and #5560

File: Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift
Comment (joshheald) [reply Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:245]: Well... I don't really have strong feelings on it. We don't _need_ the separate error, in that I've not used it and there's no specific handling of the `.unknown(_)` case. So it would be trivial to retain the `Result<String, Error>` signatures of these two methods and I'm happy to do that.

I added the `.unknown` case because I think it's useful to give some context of what the error is about in the type, while still providing the detail of the error for downstream use. We can just apply YAGNI here and I'll do that in the rework branch from #5567, and see what it looks like there.

**Review by allendav — APPROVED** [review]
We can defer `ReaderTokenProvider` and `ReaderLocationProvider` and `CardReaderConfigProvider` tuple change discussions to a separate issue.

**Pull request merged**
